eng/packaging.targets

```diff
@@ -48,8 +48,10 @@
     <_IsWindowsDesktopApp Condition="$(WindowsDesktopCoreAppLibrary.Contains('$(AssemblyName);'))">true</_IsWindowsDesktopApp>
     <_IsAspNetCoreApp Condition="$(AspNetCoreAppLibrary.Contains('$(AssemblyName);'))">true</_IsAspNetCoreApp>
     <_AssemblyInTargetingPack Condition="('$(IsNETCoreAppSrc)' == 'true' or '$(IsNetCoreAppRef)' == 'true' or '$(_IsAspNetCoreApp)' == 'true' or '$(_IsWindowsDesktopApp)' == 'true') and '$(TargetFrameworkIdentifier)' != '.NETFramework'">true</_AssemblyInTargetingPack>
-    <!-- Assembly version do not get updated in non-netfx ref pack assemblies. -->
-    <AssemblyVersion Condition="'$(_AssemblyInTargetingPack)' != 'true'">$(MajorVersion).$(MinorVersion).0.$(ServicingVersion)</AssemblyVersion>
+    <!-- The assembly version gets updated when
+      - it isn't already set (i.e. it could be pinned to a specific version for .NET Framework compatibility) and
+      - the assembly isn't part of a targeting pack (always true for .NET Framework). -->
+    <AssemblyVersion Condition="'$(AssemblyVersion)' == '' and '$(_AssemblyInTargetingPack)' != 'true'">$(MajorVersion).$(MinorVersion).0.$(ServicingVersion)</AssemblyVersion>
```

Comment:
This isn't setting `AssemblyVersion` if it's already set, but it looks like it will *always* be set, because it's done here: https://github.com/dotnet/runtime/blob/8b1d1eabe32ba781ffcce2867333dfdc53bdd635/eng/Versions.props#L18

A few other approaches:
1. A flag representing that a library controls its own AssemblyVersion
2. A property like `FixedAssemblyVersion`
3. Remove this condition entirely and just move the place where `AssemblyVersion` is hardcoded in the individual libraries to `Directory.Build.targets` after this is evaluated.

I'm inclined to do the latter.

Reply:
A few questions:

> A flag representing that a library controls its own AssemblyVersion

With this approach, we would substitute the `'$(AssemblyVersion)' == ''` condition with `'$(AssemblyControlsOwnVersion)' == 'true'`, corect? And the property would be enabled in each csproj that needs it, I assume. Would that mean we woulc have to resort to using hardcoded numbers instead of using `$(MajorVersion).$(MinorVersion).0.$(ServicingVersion)`?

> A property like FixedAssemblyVersion

Does this mean we would have to update all our OOB packages to use this new property instead of AssemblyVersion?

> and just move the place where AssemblyVersion is hardcoded in the individual libraries to Directory.Build.targets after this is evaluated.

Do you mind elaborating on this? I'm not sure I understand this option which you prefer. Do you mean having a centralized place (Directory.Build.targets) to control the OOB package versions?

Reply:
`AssemblyControlsOwnVersion` - yeah something like that.  It would only be used in these 8 projects that need to set it.  Putting it in the project would tell `packaging.targets` not to overwrite it with it.

`FixedAssemblyVersion` -- only these 8 projects.  This would just be a single property rather than `AssemblyControlsOwnVersion` and `AssemblyVersion`.  We'd then have some shared targets file that would set `AssemblyVersion` to `FixedAssemblyVersion` and it would do so in/after packaging.targets.

The third option is just moving the setting of `AssemblyVersion` out of the library-specific `Directory.Build.props` into a library-specific `Directory.Build.targets`.  Here's that change, still testing out on servicing branches: https://github.com/ericstj/runtime/commit/3a09813b43361181f47c5b15f11bc1a9e79c8711

This is all about order of evaluation and what state needs to be communicated.